src/client.rs

```diff
@@ -461,74 +562,113 @@ impl<const S: usize> ClientConnectionHandler<S> {
+    pub(crate) fn poll_close(&mut self, cx: &mut Context) -> Poll<Option<ToBehaviourEvent<S>>> {
+        if !self.closing {
+            self.closing = true;
+            self.msg.take();
+
+            if let SinkState::Ready(mut sink) = mem::replace(&mut self.sink_state, SinkState::None)
+            {
+                // Close sink but don't await for it.
+                let _ = sink.poll_close_unpin(cx);
+            }
 
-        if self.wantlist.is_none() {
-            if let Some(state) = self.sending_state.take() {
-                *state.lock().unwrap() = SendingState::Poisoned;
-                self.behaviour_waker.wake();
+            // If sending is in progress, then we don't know how much data the other end received
+            // so we consider this as "failed".
+            if matches!(
+                self.sending_state,
+                SendingState::RequestReceived(..) | SendingState::Sending(..)
+            ) {
+                self.change_sending_state(SendingState::Failed(self.connection_id));
             }
+
+            self.queue
+                .push_back(ToBehaviourEvent::ClientClosingConnection(
+                    self.peer_id,
+                    self.connection_id,
+                ));
         }
+
+        Poll::Ready(self.queue.pop_front())
     }
 
-    fn poll_outgoing(
-        &mut self,
-        cx: &mut Context,
-    ) -> Poll<
-        ConnectionHandlerEvent<ReadyUpgrade<StreamProtocol>, StreamRequester, ToBehaviourEvent<S>>,
-    > {
+    /// Each connection has its own dedicated task, which polls this method.
+    pub(crate) fn poll(&mut self, cx: &mut Context) -> Poll<ConnHandlerEvent<S>> {
         loop {
-            if self.stream_requested {
-                // We can not progress until we have a stream
-                return Poll::Pending;
+            if let Some(ev) = self.queue.pop_front() {
+                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(ev));
             }
 
-            let Some(sink) = self.sink.as_mut() else {
-                return self.poll_outgoing_no_stream();
-            };
+            if self.halted {
+                return Poll::Pending;
+            }
 
-            // Send the ongoing message before we continue to a new one
-            if ready!(sink.poll_flush_unpin(cx)).is_err() {
-                // Sink closed unexpectedly, retry
-                self.on_sink_error();
-                continue;
+            if let Some(delay) = &mut self.start_sending_timeout {
+                // If we never reached the `Sending` state within the specified
+                // time, we abort and halt this connection.
+                if delay.poll_unpin(cx).is_ready() {
+                    self.start_sending_timeout.take();
+                    self.msg.take();
+                    self.close_sink_on_error("start_sending_timeout");
+                    self.change_sending_state(SendingState::Failed(self.connection_id));
```

Comment:
Should we set sending state to failed if we never initiated sending? This will result in sending full wantlist even tho we don't have peer in the 'undefined' state. Maybe we could just close this connection and try resending on other? :thinking:

Reply:
This is intentional because `WantlistState::generate_proto_update` is stateful and alters the state each time it is called. To correctly implement what you suggest requires some kind of commit mechanism, which 
 is will have more complex code and I didn't want to implement it for the sake of simplicity.